alicloud/resource_alicloud_ons_instance.go

```diff
@@ -0,0 +1,172 @@
+		return onsClient.OnsInstanceCreate(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_ons_instance", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
+	addDebug(request.GetActionName(), raw)
+
+	var response *ons.OnsInstanceCreateResponse
+	response, _ = raw.(*ons.OnsInstanceCreateResponse)
+	d.SetId(response.Data.InstanceId)
+
+	return resourceAlicloudOnsInstanceRead(d, meta)
+}
+
+func resourceAlicloudOnsInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{client}
+
+	response, err := onsService.DescribeOnsInstance(d.Id())
+
+	if err != nil {
+		// Handle exceptions
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	d.Set("name", response.InstanceBaseInfo.InstanceName)
+	d.Set("instance_type", response.InstanceBaseInfo.InstanceType)
+	d.Set("instance_status", response.InstanceBaseInfo.InstanceStatus)
+	d.Set("release_time", time.Unix(int64(response.InstanceBaseInfo.ReleaseTime)/1000, 0).Format("2006-01-02 03:04:05"))
+
+	return nil
+}
+
+func resourceAlicloudOnsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{client}
+
+	attributeUpdate := false
+
+	request := ons.CreateOnsInstanceUpdateRequest()
+	request.InstanceId = d.Id()
+	request.PreventCache = onsService.GetPreventCache()
+
+	if d.HasChange("name") {
+		var name string
+		if v, ok := d.GetOk("name"); ok {
+			name = v.(string)
+		}
+		request.InstanceName = name
+		attributeUpdate = true
+	}
+
+	if d.HasChange("remark") {
+		var remark string
+		if v, ok := d.GetOk("remark"); ok {
+			remark = v.(string)
+		} else {
+			return fmt.Errorf("cann't change remark to empty string")
+		}
+		request.Remark = remark
+		attributeUpdate = true
+	}
+
+	if attributeUpdate {
+		raw, err := onsService.client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+			return onsClient.OnsInstanceUpdate(request)
+		})
+		if err != nil {
+			return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		addDebug(request.GetActionName(), raw)
+	}
+
+	return resourceAlicloudOnsInstanceRead(d, meta)
+}
+
+func resourceAlicloudOnsInstanceDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{client}
+
+	request := ons.CreateOnsInstanceDeleteRequest()
+	request.InstanceId = d.Id()
+	request.PreventCache = onsService.GetPreventCache()
+
+	raw, err := onsService.client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceDelete(request)
+	})
+	if err != nil {
+		if IsExceptedErrors(err, []string{OnsInstanceNotExist}) {
+			return nil
+		}
+		if IsExceptedErrors(err, []string{OnsInstanceNotEmpty}) {
+			_, err := onsService.DeleteOnsInstance(d.Id())
```

Comment:
There is no need to delete topic in the instance resource. It should be deleted in the topic resource and there should add retry for the error OnsInstanceNotEmpty, like vpc and vswitch.